File: android/app/src/main/java/com/optiswim/domain/ScoringService.kt
package com.optiswim.domain

import com.optiswim.data.model.ConditionThresholds
import com.optiswim.data.model.FactorWeights
import com.optiswim.data.model.MarineConditions
import com.optiswim.data.model.ScoreRating
import com.optiswim.data.model.SwimScore
import com.optiswim.data.model.SwimmerLevel
import com.optiswim.data.model.TidePreference

object DefaultProfiles {
    fun thresholds(level: SwimmerLevel): ConditionThresholds = when (level) {
        SwimmerLevel.BEGINNER -> ConditionThresholds(20.0, 0.3, 15.0, 20.0, TidePreference.SLACK_OR_HIGH, false)
        SwimmerLevel.INTERMEDIATE -> ConditionThresholds(16.0, 0.6, 20.0, 30.0, TidePreference.ANY, true)
        SwimmerLevel.EXPERIENCED -> ConditionThresholds(12.0, 1.0, 30.0, 45.0, TidePreference.ANY, true)
        SwimmerLevel.COLD_WATER -> ConditionThresholds(8.0, 1.0, 30.0, 45.0, TidePreference.ANY, true)
    }

    fun weights(level: SwimmerLevel): FactorWeights = when (level) {
        SwimmerLevel.BEGINNER -> FactorWeights(0.30, 0.25, 0.20, 0.10, 0.10, 0.05)
        SwimmerLevel.INTERMEDIATE -> FactorWeights(0.25, 0.25, 0.20, 0.10, 0.10, 0.10)
        SwimmerLevel.EXPERIENCED -> FactorWeights(0.15, 0.25, 0.20, 0.10, 0.15, 0.15)
        SwimmerLevel.COLD_WATER -> FactorWeights(0.10, 0.30, 0.25, 0.10, 0.10, 0.15)
    }
}

object ScoringService {
    private const val ABSOLUTE_MIN_TEMP = 5.0
    private const val ABSOLUTE_MAX_WAVE = 2.0
    private const val ABSOLUTE_MAX_WIND = 50.0

    fun calculateScore(conditions: MarineConditions, level: SwimmerLevel): SwimScore {
        val thresholds = DefaultProfiles.thresholds(level)
        val weights = DefaultProfiles.weights(level)

        if (conditions.waveHeight > ABSOLUTE_MAX_WAVE) {
            return SwimScore(0.0, ScoreRating.Dangerous, listOf("Dangerous wave conditions"))
        }

        if (conditions.windSpeed > ABSOLUTE_MAX_WIND) {
            return SwimScore(0.0, ScoreRating.Dangerous, listOf("Dangerous wind conditions"))
        }

        val tempScore = tempScore(conditions.seaSurfaceTemperature, thresholds)
        val waveScore = waveScore(conditions.waveHeight, thresholds)
        val windScore = windScore(conditions.windSpeed, conditions.windGusts, thresholds)
        val directionScore = if (isOnshore(conditions.windDirection) && !thresholds.acceptOnshoreWind) 0.5 else 1.0
        val weatherScore = weatherScore(conditions.weatherCode)
        val tideScore = tideScore(resolveTideState(conditions), thresholds.preferredTide)

        val weighted = (tempScore * weights.temperature) +
            (waveScore * weights.wave) +
            (windScore * weights.wind) +
            (directionScore * weights.direction) +
            (weatherScore * weights.weather) +
            (tideScore * weights.tide)

        val finalScore = (weighted * 100).coerceIn(0.0, 100.0)
        val rating = when {
            finalScore >= 85 -> ScoreRating.Excellent
            finalScore >= 70 -> ScoreRating.Good
            finalScore >= 55 -> ScoreRating.Fair
            finalScore >= 40 -> ScoreRating.Poor
            else -> ScoreRating.Dangerous
        }

        return SwimScore(finalScore, rating, warnings(conditions, thresholds))
    }

    private fun tempScore(temp: Double, thresholds: ConditionThresholds): Double = when {
        temp >= 22 && temp <= 25 -> 1.0
        temp >= thresholds.minWaterTemp -> (1.0 - (kotlin.math.abs(temp - 23.5) * 0.05)).coerceAtLeast(0.6)
        temp >= ABSOLUTE_MIN_TEMP -> ((temp - ABSOLUTE_MIN_TEMP) / (thresholds.minWaterTemp - ABSOLUTE_MIN_TEMP) * 0.5).coerceAtLeast(0.1)
        else -> 0.0
    }

    private fun waveScore(height: Double, thresholds: ConditionThresholds): Double = when {
        height <= 0.2 -> 1.0
        height <= thresholds.maxWaveHeight -> (1.0 - (height - 0.2) / (thresholds.maxWaveHeight - 0.2)).coerceAtLeast(0.6)
        height <= ABSOLUTE_MAX_WAVE -> ((1.0 - (height - thresholds.maxWaveHeight) / (ABSOLUTE_MAX_WAVE - thresholds.maxWaveHeight)) * 0.5).coerceAtLeast(0.1)
        else -> 0.0
    }

    private fun windScore(speed: Double, gusts: Double, thresholds: ConditionThresholds): Double {
        var score = when {
            speed <= 10 -> 1.0
            speed <= thresholds.maxWindSpeed -> (1.0 - (speed - 10) / (thresholds.maxWindSpeed - 10)).coerceAtLeast(0.6)
            speed <= ABSOLUTE_MAX_WIND -> ((1.0 - (speed - thresholds.maxWindSpeed) / (ABSOLUTE_MAX_WIND - thresholds.maxWindSpeed)) * 0.5).coerceAtLeast(0.1)
            else -> 0.0
        }

        if (gusts > thresholds.maxWindGusts) {
            score *= 0.7
        } else if (gusts > speed * 1.5) {
            score *= 0.85
        }

        return score
    }

    private fun weatherScore(code: Int): Double = when (code) {
        0 -> 1.0
        1 -> 0.95
        2 -> 0.9
        3 -> 0.75
        51 -> 0.6
        53, 55 -> 0.4
        45, 48 -> 0.3
        else -> 0.2
    }

    private fun tideScore(state: String, preference: TidePreference): Double {
        val normalized = if (state == "rising" || state == "falling") "mid" else state
        return when (preference) {
            TidePreference.ANY -> 1.0
            TidePreference.HIGH -> if (normalized == "high") 1.0 else if (normalized == "mid") 0.7 else 0.5
            TidePreference.LOW -> if (normalized == "low") 1.0 else if (normalized == "mid") 0.7 else 0.5
            TidePreference.SLACK_OR_HIGH -> if (normalized == "high" || normalized == "mid") 1.0 else 0.6
        }
    }

    private fun resolveTideState(conditions: MarineConditions): String {
        conditions.tidePhase?.let { return it.lowercase() }
        return when {
            conditions.seaLevelHeight > 0.5 -> "high"
            conditions.seaLevelHeight < -0.5 -> "low"
            else -> "mid"
        }
    }

    private fun warnings(conditions: MarineConditions, thresholds: ConditionThresholds): List<String> {
        val warnings = mutableListOf<String>()
        if (conditions.seaSurfaceTemperature < 15) warnings.add("Cold water shock risk")
        if (conditions.waveHeight > thresholds.maxWaveHeight * 0.8) warnings.add("Higher waves expected")
        if (conditions.windSpeed > thresholds.maxWindSpeed * 0.8) warnings.add("Wind approaching limit")
        if (conditions.windGusts > conditions.windSpeed * 1.5) warnings.add("Gusty conditions")
        if (conditions.uvIndex >= 6) warnings.add("High UV index")
        return warnings
    }

    private fun isOnshore(direction: Double): Boolean {
        return direction in 180.0..360.0
    }
}

File: OptiSwim/Services/ScoringService.swift
import Foundation

// MARK: - Scoring Service

struct ScoringService {
    
    // MARK: - Calculate Swim Score
    
    static func calculateScore(
        conditions: MarineConditions,
        profile: UserProfile
    ) -> SwimScore {
        let thresholds = profile.thresholds
        let weights = profile.weights
        
        var warnings: [SafetyWarning] = []
        
        // 1. Check hard safety limits (instant disqualification)
        if conditions.waveHeight > ConditionThresholds.absoluteMaxWave {
            return SwimScore(
                value: 0,
                rating: .dangerous,
                warnings: [.dangerousWaves],
                breakdown: zeroBreakdown(),
                optimalWindow: nil
            )
        }
        
        if conditions.windSpeed > ConditionThresholds.absoluteMaxWind {
            return SwimScore(
                value: 0,
                rating: .dangerous,
                warnings: [.dangerousWind],
                breakdown: zeroBreakdown(),
                optimalWindow: nil
            )
        }
        
        let isSafe = WeatherCode(rawValue: conditions.weatherCode)?.isSafeForSwimming ?? true
        if !isSafe {
            if conditions.weatherCode >= 95 { // Thunderstorm
                return SwimScore(
                    value: 0,
                    rating: .dangerous,
                    warnings: [.stormWarning],
                    breakdown: zeroBreakdown(),
                    optimalWindow: nil
                )
            }
        }
        
        // 2. Calculate individual factor scores
        let tempScore = calculateTempScore(conditions.waterTemperature, thresholds: thresholds)
        let waveScore = calculateWaveScore(conditions.waveHeight, thresholds: thresholds)
        let windScore = calculateWindScore(conditions.windSpeed, gusts: conditions.windGusts, thresholds: thresholds)
        let directionScore = calculateDirectionScore(conditions.windDirection, acceptOnshore: thresholds.acceptOnshoreWind)
        let weatherScore = calculateWeatherScore(conditions.weatherCode)
        let tideScore = calculateTideScore(conditions.tideState, preference: thresholds.preferredTide)
        
        // 3. Apply weights
        let weightedScore = (tempScore * weights.temperature) +
                           (waveScore * weights.wave) +
                           (windScore * weights.wind) +
                           (directionScore * weights.direction) +
                           (weatherScore * weights.weather) +
                           (tideScore * weights.tide)
        
        // Normalize to 0-100
        let finalScore = min(100, max(0, weightedScore * 100))
        
        // 4. Generate warnings
        warnings = generateWarnings(conditions: conditions, thresholds: thresholds)
        
        let breakdown = ScoreBreakdown(
            temperatureScore: tempScore * 100,
            waveScore: waveScore * 100,
            windScore: windScore * 100,
            directionScore: directionScore * 100,
            weatherScore: weatherScore * 100,
            tideScore: tideScore * 100
        )
        
        return SwimScore(
            value: finalScore,
            rating: ScoreRating.from(score: finalScore),
            warnings: warnings,
            breakdown: breakdown,
            optimalWindow: nil
        )
    }
    
    // MARK: - Find Optimal Window
    
    static func findOptimalWindow(
        forecast: [HourlyForecast],
        profile: UserProfile,
        minDuration: TimeInterval = 7200 // 2 hours
    ) -> TimeWindow? {
        guard forecast.count >= 2 else { return nil }
        
        // Score each hour
        let scores = forecast.map { hour -> (Date, Double) in
            let score = calculateScore(conditions: hour.conditions, profile: profile)
            return (hour.timestamp, score.value)
        }
        
        // Find the best consecutive window
        var bestWindow: TimeWindow?
        var bestAverageScore: Double = 0
        
        let minHours = Int(minDuration / 3600)
        
        for i in 0..<(scores.count - minHours + 1) {
            let windowScores = scores[i..<(i + minHours)]
            let averageScore = windowScores.map { $0.1 }.reduce(0, +) / Double(minHours)
            
            // Only consider windows with good scores (>= 60)
            if averageScore >= 60 && averageScore > bestAverageScore {
                bestAverageScore = averageScore
                bestWindow = TimeWindow(
                    start: windowScores.first!.0,
                    end: windowScores.last!.0.addingTimeInterval(3600),
                    averageScore: averageScore
                )
            }
        }
        
        return bestWindow
    }
    
    // MARK: - Individual Score Calculations
    
    private static func calculateTempScore(_ temp: Double, thresholds: ConditionThresholds) -> Double {
        // Perfect: 22-25°C
        // Good: within user threshold
        // Poor: approaching cold water shock
        
        if temp >= 22 && temp <= 25 {
            return 1.0
        } else if temp >= thresholds.minWaterTemp {
            let distance = min(abs(temp - 22), abs(temp - 25))
            return max(0.6, 1.0 - (distance * 0.05))
        } else if temp >= ConditionThresholds.absoluteMinTemp {
            let ratio = (temp - ConditionThresholds.absoluteMinTemp) / (thresholds.minWaterTemp - ConditionThresholds.absoluteMinTemp)
            return max(0.1, ratio * 0.5)
        } else {
            return 0.0
        }
    }
    
    private static func calculateWaveScore(_ height: Double, thresholds: ConditionThresholds) -> Double {
        // Perfect: < 0.2m
        // Good: within threshold
        // Poor: approaching dangerous
        
        if height <= 0.2 {
            return 1.0
        } else if height <= thresholds.maxWaveHeight {
            let ratio = 1.0 - (height - 0.2) / (thresholds.maxWaveHeight - 0.2)
            return max(0.6, ratio)
        } else if height <= ConditionThresholds.absoluteMaxWave {
            let ratio = 1.0 - (height - thresholds.maxWaveHeight) / (ConditionThresholds.absoluteMaxWave - thresholds.maxWaveHeight)
            return max(0.1, ratio * 0.5)
        } else {
            return 0.0
        }
    }
    
    private static func calculateWindScore(_ speed: Double, gusts: Double, thresholds: ConditionThresholds) -> Double {
        // Perfect: < 10 km/h
        // Good: within threshold
        // Poor: gusty or approaching dangerous
        
        var score: Double = 1.0
        
        if speed <= 10 {
            score = 1.0
        } else if speed <= thresholds.maxWindSpeed {
            let ratio = 1.0 - (speed - 10) / (thresholds.maxWindSpeed - 10)
            score = max(0.6, ratio)
        } else if speed <= ConditionThresholds.absoluteMaxWind {
            let ratio = 1.0 - (speed - thresholds.maxWindSpeed) / (ConditionThresholds.absoluteMaxWind - thresholds.maxWindSpeed)
            score = max(0.1, ratio * 0.5)
        } else {
            score = 0.0
        }
        
        // Penalize gusty conditions
        if gusts > thresholds.maxWindGusts {
            score *= 0.7
        } else if gusts > speed * 1.5 {
            score *= 0.85
        }
        
        return score
    }
    
    private static func calculateDirectionScore(_ direction: Double, acceptOnshore: Bool) -> Double {
        // This is simplified - in reality you'd need coast orientation
        // For now, assume onshore is from 180-360° (south to north through west)
        
        let isOnshore = direction >= 180 && direction <= 360
        
        if isOnshore && !acceptOnshore {
            return 0.5
        }
        
        return 1.0
    }
    
    private static func calculateWeatherScore(_ code: Int) -> Double {
        guard let weather = WeatherCode(rawValue: code) else {
            return 0.5
        }
        
        switch weather {
        case .clearSky:
            return 1.0
        case .mainlyClear:
            return 0.95
        case .partlyCloudy:
            return 0.9
        case .overcast:
            return 0.75
        case .drizzleLight:
            return 0.6
        case .drizzleModerate, .drizzleDense:
            return 0.4
        case .fog, .depositingFog:
            return 0.3
        default:
            return 0.2
        }
    }
    
    private static func calculateTideScore(_ state: TideState, preference: TidePreference) -> Double {
        let normalizedState: TideState = (state == .rising || state == .falling) ? .mid : state
        switch preference {
        case .any:
            return 1.0
        case .high:
            return normalizedState == .high ? 1.0 : (normalizedState == .mid ? 0.7 : 0.5)
        case .low:
            return normalizedState == .low ? 1.0 : (normalizedState == .mid ? 0.7 : 0.5)
        case .slackOrHigh:
            return (normalizedState == .high || normalizedState == .mid) ? 1.0 : 0.6
        }
    }
    
    // MARK: - Generate Warnings
    
    private static func generateWarnings(conditions: MarineConditions, thresholds: ConditionThresholds) -> [SafetyWarning] {
        var warnings: [SafetyWarning] = []
        
        // Cold water warning
        if conditions.waterTemperature < 15 {
            warnings.append(.coldWaterShock)
        }
        
        // Wave warnings
        if conditions.waveHeight > thresholds.maxWaveHeight * 0.8 {
            warnings.append(.dangerousWaves)
        }
        
        // Wind warnings
        if conditions.windSpeed > thresholds.maxWindSpeed * 0.8 {
            warnings.append(.dangerousWind)
        }
        
        if conditions.windGusts > conditions.windSpeed * 1.5 {
            warnings.append(.gustyConditions)
        }
        
        // UV warning
        if conditions.uvIndex >= 6 {
            warnings.append(.highUV)
        }
        
        // Visibility warning
        if let weather = WeatherCode(rawValue: conditions.weatherCode),
           weather == .fog || weather == .depositingFog {
            warnings.append(.lowVisibility)
        }
        
        // Onshore wind warning
        let isOnshore = conditions.windDirection >= 180 && conditions.windDirection <= 360
        if isOnshore && !thresholds.acceptOnshoreWind && conditions.windSpeed > 10 {
            warnings.append(.onshoreWind)
        }
        
        return warnings
    }
    
    private static func zeroBreakdown() -> ScoreBreakdown {
        ScoreBreakdown(
            temperatureScore: 0,
            waveScore: 0,
            windScore: 0,
            directionScore: 0,
            weatherScore: 0,
            tideScore: 0
        )
    }
}
